### PR TITLE
Major refactoring of usercard component

### PR DESCRIPTION
**Pull request by freddidierRTE**
The objective is to externalize the business logic outside of the angular component

Have a look to the comments in userCard.view.ts for explanation on the structure of the business code

To be noted :
 - this commit solve ExpressionChangedAfterItHasBeenCheckedError
 - we have now a unique message when trying to access usercard with a config that do not permit to send card
 - the commit remove use of emitter in code base and use publisher to be consistant

After this refactoring , there is still business code in angular component regarding recipient preview , it may be refactor in another issue

Nothing in release note

### CODE REVIEW

**Comment by vlo-rte** [comment]
When I start opfab and then I go (for the first time) on the usercard menu, I see this error in the console : 
image

**Comment by vlo-rte** [comment]
When I create a usercard, then I open it, then I click "create a copy", I see these two errors : 
image

**Comment by vlo-rte** [comment]
When I click "edit card" with the same card than previously, I see this error : 
image

File: ui/main/src/app/business/view/userCard/userCard.view.ts [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+or reinitialized  after the template is rendered. That is one of the reasons why we have the two methods
+initFieldsThatNeedToBeSetBeforeExecutingTemplateScripts and initFieldsThatNeedToBeSetAfterExecutingTemplateScripts.
+
+Another important point is the asynchronous nature of the rendering of the template, we need to wait
+for the template rendering and the template scripts execution to complete before calling the next method.
+
+The CardBuilder class is used both to prepare the card and child card to be sent and to get the card with child cards
+for preview.
+
+The CardSend is used to send the card and child card to the backend after the card has been build via the CardBuilder
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.ts]: build => built

File: ui/main/src/app/business/view/userCard/userCard.view.ts [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+and when user click on the send button.
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.ts]: I suggest : "and when the user clicks" ...

**Comment by vlo-rte** [comment]
image
If it is possible, I suggest to set void for all the functions or to remove it from all of them, so that the code is homogeneous ?

File: build.gradle [hunk 3]
```diff
@@ -118,6 +118,12 @@ sonar {
         property 'sonar.exclusions','**/generated/**/*,**/src/docs/**/*,**/*Application.java,**/node_modules/**/*'
         // exclusions for analyse of duplicated code , test classes are excluded 
         property 'sonar.cpd.exclusions','**/model/**/*,**/src/test/**/*,**/*.spec.ts,src/test/**,node-services/**/tests/*'
+        
+        // ignore rule S2004 "Refactor this code to not nest functions more than 4 levels deep" for test files
+        // we need to create nested test and this rules is blocking us
```
Comment (vlo-rte) [inline build.gradle]: rules=>rule

File: ui/main/src/app/business/view/userCard/cardSender/cardSender.spec.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            cardServerMock.setResponseFunctionForPostCard(() => {
+                return new ServerResponse(null, ServerResponseStatus.NOT_FOUND, null);
+            });
+            const alertMessageReceiver = new AlertMessageReceiver();
+            await cardSender.sendCardAndChildCard(card);
+            const alertMessage = await alertMessageReceiver.getMessageReceived();
+            expect(alertMessage.i18n.key).toEqual('userCard.error.impossibleToSendCard');
+            expect(alertMessage.level).toEqual(MessageLevel.ERROR);
+        });
+        it('Should set the card id as last card send for soundNotification service to not play sound for the card', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardSender/cardSender.spec.ts]: send=>sent

File: ui/main/src/app/business/view/userCard/cardSender/cardSender.spec.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            await cardSender.sendCardAndChildCard(card);
+            expect(SoundNotificationService.getLastSentCardId()).toEqual(card.process + '.' + card.processInstanceId);
+        });
+
+        it('Should set the card id as last card send for systemNotification service to not play sound for the card', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardSender/cardSender.spec.ts]: send=>sent

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+            jasmine.clock().mockDate(new Date(50));
+        });
+        afterEach(() => {
+            jasmine.clock().uninstall();
+        });
+        it('should display error if template does not return specific card information', async () => {
+            opfab.currentUserCard.registerFunctionToGetSpecificCardInformation(() => null);
+            await checkAlertMessage('userCard.error.templateError');
+        });
+        it('should display error message provide by template if specific information is not valid', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: provide=>provided

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+    });
+    describe('Severity', async () => {
+        it('should be selected by user if severity visible', async () => {
+            setSpecificCardInformation({valid: true, card: {severity: 'ACTION'}});
+            cardBuilder.setFieldVisible(InputFieldName.Severity, true);
+            cardBuilder.setSeveritySelectedByUser(Severity.ALARM);
+            const card = await cardBuilder.getCard();
+            expect(card.severity).toBe(Severity.ALARM);
+        });
+        it('should be severity provide by template if severity not visible', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: provide=>provided

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+            setSpecificCardInformation({valid: true, card: {severity: 'ACTION'}});
+            cardBuilder.setFieldVisible(InputFieldName.Severity, false);
+            const card = await cardBuilder.getCard();
+            expect(card.severity).toBe(Severity.ACTION);
+        });
+        it('Should be INFORMATION if severity not visible and template not provide severity', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: I suggest : "and template does not provide severity"

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+            expect(card.entityRecipients).toEqual(['ENTITY3', 'ENTITY4']);
+        });
+        it('Should be the union of recipients from template and recipients form user if field is visible', async () => {
+            setSpecificCardInformation({valid: true, card: {entityRecipients: ['ENTITY3', 'ENTITY4']}});
+            cardBuilder.setFieldVisible(InputFieldName.Recipients, true);
+            cardBuilder.setRecipientsSelectedByUser(['ENTITY1', 'ENTITY2', 'ENTITY3']);
+            const card = await cardBuilder.getCard();
+            expect(card.entityRecipients).toEqual(['ENTITY1', 'ENTITY2', 'ENTITY3', 'ENTITY4']);
+        });
+        it('Should be empty if not visible and template not provide it', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: I suggest : "and template does not provide it"

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+    });
+    describe('RecipientsForInformation', () => {
+        it('Should be set from template if not visible', async () => {
+            setSpecificCardInformation({valid: true, card: {entityRecipientsForInformation: ['ENTITY1', 'ENTITY2']}});
+            cardBuilder.setFieldVisible(InputFieldName.RecipientsForInformation, false);
+            cardBuilder.setRecipientsForInformationSelectedByUser([]);
+            const card = await cardBuilder.getCard();
+            expect(card.entityRecipientsForInformation).toEqual(['ENTITY1', 'ENTITY2']);
+        });
+        it('Should be the union of recipientsForInformation from template and recipientsFromInformation form user if field is visible', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: form=>from

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+            expect(card.entityRecipientsForInformation).toEqual(['ENTITY1', 'ENTITY2', 'ENTITY3', 'ENTITY4']);
+        });
+        it('Should remove from recipientForInformation set by template, recipients selected by user ', async () => {
+            setSpecificCardInformation({valid: true, card: {entityRecipientsForInformation: ['ENTITY2', 'ENTITY3']}});
+            cardBuilder.setFieldVisible(InputFieldName.Recipients, true);
+            cardBuilder.setRecipientsSelectedByUser(['ENTITY1', 'ENTITY2']);
+            const card = await cardBuilder.getCard();
+            expect(card.entityRecipientsForInformation).toEqual(['ENTITY3']);
+        });
+        it('Should remove from recipientForInformation selected by user, recipients selected by user ', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: recipientForInformation=>recipientsForInformation

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+            expect(card.entityRecipientsForInformation).toEqual(['ENTITY1', 'ENTITY2']);
+        });
+        it('Should be the union of recipientsForInformation from template and recipientsFromInformation form user if field is visible', async () => {
+            setSpecificCardInformation({valid: true, card: {entityRecipientsForInformation: ['ENTITY3', 'ENTITY4']}});
+            cardBuilder.setFieldVisible(InputFieldName.RecipientsForInformation, true);
+            cardBuilder.setRecipientsForInformationSelectedByUser(['ENTITY1', 'ENTITY2', 'ENTITY3']);
+            const card = await cardBuilder.getCard();
+            expect(card.entityRecipientsForInformation).toEqual(['ENTITY1', 'ENTITY2', 'ENTITY3', 'ENTITY4']);
+        });
+        it('Should remove from recipientForInformation set by template, recipients selected by user ', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: recipientForInformation=>recipientsForInformation

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+                });
+                it('Should be provided by template if not visible', async () => {
+                    setSpecificCardInformation({valid: true, card: {[field]: 300}});
+                    cardBuilder.setFieldVisible(InputFieldName.StartDate, true);
+                    cardBuilder.setStartDate(150);
+                    cardBuilder.setFieldVisible(field, false);
+                    const card = await cardBuilder.getCard();
+                    expect(card[field]).toBe(300);
+                });
+                it('Should be undefined if not visible and template not provide it', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: I suggest "and template does not provide it"

File: ui/main/src/app/business/view/userCard/cardSender/cardSender.spec.ts [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+        const childCard = getOneCard();
+
+        it('Should send child card to the back end', async () => {
+            await cardSender.sendCardAndChildCard(card, childCard);
+            expect(cardServerMock.cardsPosted.length).toBe(2);
+            expect(cardServerMock.cardsPosted[1].data).toEqual(childCard.data);
+            expect(cardServerMock.cardsPosted[1].publisher).toEqual(childCard.publisher);
+            expect(cardServerMock.cardsPosted[1].process).toEqual(childCard.process);
+        });
+        it('Child card should contains child card + cards id and card uid of the parent created card', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardSender/cardSender.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/cardSender/cardSender.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+            message: '',
+            level: MessageLevel.ERROR,
+            i18n: new I18n('userCard.error.impossibleToSendCard')
+        });
+    }
+    private async sendChildCard(
+        childCard: Card,
+        responseFromCardPost: ServerResponse<CardCreationReportData>,
+        setCurrentDateForStartDate
+    ) {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/cardSender/cardSender.ts]: I suggest to write the return type if it is possible.

File: ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.spec.ts [hunk 16]
```diff
@@ -0,0 +1,670 @@
...
+                new ComputedPerimeter('process1', 'state1_0', RightsEnum.ReceiveAndWrite)
+            ]);
+            initView();
+            userCardUIControlMock.initMock();
+            processStateForm.userClickOnProcess('process0');
+            expect(userCardUIControlMock.states).toBeUndefined();
+            expect(userCardUIControlMock.inputVisibility_FctCalls[InputFieldName.State]).toEqual(undefined);
+        });
+    });
+    describe('User click on processGroup', () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.spec.ts]: click=>clicks

File: ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.ts [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+
+        const processNotInProcessGroups = this.getProcessesForUserCardNotInProcessGroupsForProcessList(processList);
+        if (processNotInProcessGroups.length > 0) {
+            processGroupList.push(new ProcessGroupForUserCard('--', '--', processNotInProcessGroups));
+        }
+        this.sortListByLabel(processGroupList);
+        return processGroupList;
+    }
+
+    private checkIfUserHasAtLeastOneEntityAllowedToSendCard(processId: string, stateId: string) {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.ts]: I suggest to write the return type

File: ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.ts [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+            return entityIds.some((entityId) => userEntities.includes(entityId));
+        }
+        return true;
+    }
+
+    private sortListByLabel(list: {label: string}[]): void {
+        list.sort((a, b) => a.label.localeCompare(b.label));
+    }
+
+    private getProcessesForUserCardNotInProcessGroupsForProcessList(processes: ProcessForUserCard[]) {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.ts]: I suggest to write the return type

File: ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.ts [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+
+    public hasAtLeastOneStateAllowedToSendCard(): boolean {
+        return this.selectedStateId !== undefined;
+    }
+
+    public getSelectedProcessGroupId(): string {
+        return this.selectedProcessGroupId;
+    }
+
+    public userClickOnState(stateId: string) {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/processStatesForm/processStatesForm.ts]: userClickOnState=>userClicksOnState ?

File: ui/main/src/app/business/view/userCard/publisherForm/publisherForm.spec.ts [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+            it('Should not select publisher of existing card in copy mode if user is member of the publishing entity', async () => {
+                await setUserWithEntities(['ENTITY1', 'ENTITY2', 'ENTITY4']);
+                await setProcessConfigWithUserCardConfig({});
+                const card = getOneCard({publisher: 'ENTITY4'});
+                publisherForm.setProcessAndState('testProcessId', 'testStateId', card, EditionMode.COPY);
+                expect(publisherForm.getSelectedPublisher()).toEqual('ENTITY1');
+            });
+        });
+    });
+    describe('User select publisher on UI', () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/publisherForm/publisherForm.spec.ts]: select=>selects

File: ui/main/src/app/business/view/userCard/publisherForm/publisherForm.ts [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+        const publishers = this.buildPublisherMultiselectList(entitiesAllowedToSendCard);
+        const initialSelectedPublisher = this.getInitialSelectedPublisher(publishers, card);
+        if (isPublisherVisible) {
+            this.userCardUIControl.setPublisherList(publishers, initialSelectedPublisher);
+        }
+        this.selectPublisher(initialSelectedPublisher);
+    }
+
+    private getUserEntitiesWithCardSenderRole() {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/publisherForm/publisherForm.ts]: I suggest to write the return type of the function

File: ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts [hunk 22]
```diff
@@ -0,0 +1,402 @@
...
+            });
+            it(`Should be set to true if recipient visibility is not defined in state configuration`, async () => {
+                await setProcessConfigWithUserCardConfig({});
+                recipientsForm.setProcessAndState('process1', 'state1_1');
+                expect(recipientsForm.isRecipientVisible()).toEqual(true);
+                expect(userCardUIControl.inputVisibility_FctCalls['recipients']).toEqual(true);
+            });
+        });
+        describe('Option list', () => {
+            it(`Should contains all entities with Role CARD_RECEIVER`, async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts [hunk 23]
```diff
@@ -0,0 +1,402 @@
...
+                await setProcessConfigWithUserCardConfig({});
+                recipientsForm = new RecipientsForm(userCardUIControl);
+                recipientsForm.setProcessAndState('process1', 'state1_1');
+                expect(userCardUIControl.recipients).toEqual([
+                    {id: 'ENTITY1', label: 'ENTITY1_DESCRIPTION'},
+                    {id: 'ENTITY2', label: 'ENTITY2_DESCRIPTION'},
+                    {id: 'PARENT_ENTITY', label: 'PARENT_ENTITY_DESCRIPTION'}
+                ]);
+            });
+            it(`Should contains all entities by label alphabetical order`, async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts [hunk 24]
```diff
@@ -0,0 +1,402 @@
...
+            });
+            it(`Should be set to false  if recipientForInformation visibility is not defined in state configuration`, async () => {
+                await setProcessConfigWithUserCardConfig({});
+                recipientsForm.setProcessAndState('process1', 'state1_1');
+                expect(recipientsForm.isRecipientForInformationVisible()).toEqual(false);
+                expect(userCardUIControl.inputVisibility_FctCalls['recipientsForInformation']).toEqual(false);
+            });
+        });
+        describe('Option list', () => {
+            it(`Should contains all entities with Role CARD_RECEIVER`, async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts [hunk 25]
```diff
@@ -0,0 +1,402 @@
...
+                await setProcessConfigWithUserCardConfig({});
+                recipientsForm = new RecipientsForm(userCardUIControl);
+                recipientsForm.setProcessAndState('process1', 'state1_1');
+                expect(userCardUIControl.recipientsForInformation).toEqual([
+                    {id: 'ENTITY1', label: 'ENTITY1_DESCRIPTION'},
+                    {id: 'ENTITY2', label: 'ENTITY2_DESCRIPTION'},
+                    {id: 'PARENT_ENTITY', label: 'PARENT_ENTITY_DESCRIPTION'}
+                ]);
+            });
+            it(`Should contains all entities by label alphabetical order`, async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/recipientsForm/recipientsForm.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts [hunk 26]
```diff
@@ -0,0 +1,849 @@
...
+    function setCardDataResponse(cardData: CardData) {
+        cardServerMock.setResponseFunctionForLoadCard(() => {
+            return new ServerResponse(cardData, ServerResponseStatus.OK, '');
+        });
+    }
+
+    async function setCurrentUserConfiguration(
+        computedPerimeters: ComputedPerimeter[],
+        permissions: Array<PermissionEnum> = null
+    ): Promise<void> {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.spec.ts:52]: There is no return instruction in this function so maybe the return type should be void (or nothing of course) ?

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts [hunk 27]
```diff
@@ -0,0 +1,849 @@
...
+                    childCards: []
+                },
+                EditionMode.EDITION
+            );
+            setSpecificCardInformation({valid: true, card: {data: 'test'}});
+            await userCardView.prepareCardToSend();
+            const childCards = userCardView.getCardWithChildCardsForPreview().childCards;
+            expect(childCards.length).toEqual(0);
+        });
+        it('Should contains child cards of existing card to edit if keepChildCard = true (deprecated feature)', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts [hunk 28]
```diff
@@ -0,0 +1,849 @@
...
+                EditionMode.EDITION
+            );
+            setSpecificCardInformation({valid: true, card: {data: 'test'}});
+            await userCardView.prepareCardToSend();
+            const childCards = userCardView.getCardWithChildCardsForPreview().childCards;
+            expect(childCards.length).toEqual(1);
+            expect(childCards[0].process).toEqual('process1');
+            expect(childCards[0].state).toEqual('state1_1');
+        });
+        it('Should not contains child cards of existing card to edit if keepChildCard = false (deprecated feature)', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts [hunk 29]
```diff
@@ -0,0 +1,849 @@
...
+                    childCards: [getOneCard({process: 'process1', state: 'state1_1'})]
+                },
+                EditionMode.EDITION
+            );
+            setSpecificCardInformation({valid: true, card: {data: 'test'}});
+            await userCardView.prepareCardToSend();
+            const childCards = userCardView.getCardWithChildCardsForPreview().childCards;
+            expect(childCards.length).toEqual(0);
+        });
+        it('Should contains child cards of existing card to edit if card action contains KEEP_CHILD_CARD', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts [hunk 30]
```diff
@@ -0,0 +1,849 @@
...
+                summary: new I18n('mySummary')
+            });
+            childCard.publisher = undefined;
+            setSpecificCardInformation({valid: true, card: {data: 'test'}, childCard: childCard});
+            await userCardView.prepareCardToSend();
+            const childCards = userCardView.getCardWithChildCardsForPreview().childCards;
+            expect(childCards.length).toEqual(1);
+            expect(childCards[0].data).toEqual('childData');
+        });
+        it('Should contains child card from specific card information + child card from existing card if card action contains KEEP_CHILD_CARDS', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts [hunk 31]
```diff
@@ -0,0 +1,849 @@
...
+                    childCards: [getOneCard({process: 'process1', state: 'state1_1'})]
+                },
+                EditionMode.COPY
+            );
+            setSpecificCardInformation({valid: true, card: {data: 'test'}});
+            await userCardView.prepareCardToSend();
+            const childCards = userCardView.getCardWithChildCardsForPreview().childCards;
+            expect(childCards.length).toEqual(0);
+        });
+        it('Should contains child card from specific card information', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.spec.ts]: contains=>contain

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts [hunk 32]
```diff
@@ -0,0 +1,849 @@
...
+                summary: new I18n('mySummary')
+            });
+            childCard.publisher = undefined;
+            setSpecificCardInformation({valid: true, card: {data: 'test'}, childCard: childCard});
+            await userCardView.prepareCardToSend();
+            await userCardView.sendCardAncChildCard();
+            expect(cardServerMock.cardsPosted[0].data).toEqual('test');
+            expect(cardServerMock.cardsPosted[1].data).toEqual('childData');
+        });
+        it('When startDate is set to currentDate it should send card and childCard with the date when user click on send and not when preview has been clicked ', async () => {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.spec.ts]: click=>clicks

File: ui/main/src/app/business/view/userCard/userCard.view.ts [hunk 33]
```diff
@@ -0,0 +1,267 @@
...
+    public userSetPublisher(publisherEntityId: string) {
+        this.publisherForm.userSelectsPublisher(publisherEntityId);
+    }
+    public userSetRecipients(entityRecipientIds: string[]) {
+        this.recipientsForm.setSelectedRecipients([...entityRecipientIds]);
+    }
+    public userSetRecipientsForInformation(entityRecipientsForInformationIds: string[]) {
+        this.recipientsForm.setSelectedRecipientsForInformation([...entityRecipientsForInformationIds]);
+    }
+    public async prepareCardToSend(): Promise<void> {
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.ts:209]: I think the return type of the function should be void (or nothing)

File: ui/main/src/app/modules/usercard/usercard.component.html [hunk 34]
```diff
@@ -1,98 +1,165 @@
...
+            <div [hidden]="templateLoadingInProgress" class="justify-content-center"
+                style="display:flex;width:100%;align-content:center;flex-wrap: wrap;">
+                <div style="width:280px; margin: 30px" class="nopaddingrow opfab-input"
+                    id="opfab-usercard-input-startDate" hidden>
+                    <label for="opfab-usercard-startdate-choice-input" translate>userCard.filters.startDate</label>
+                    <input id="opfab-usercard-startdate-choice-input" type="datetime-local" formControlName="startDate"
+                        (change)="onStartDateChange()">
+                </div>
+
+                <div style="width: 280px; margin: 30px" class="nopaddingrow opfab-input"
```
Comment (vlo-rte) [inline ui/main/src/app/modules/usercard/usercard.component.html]: I think the css class nopaddingrow has been removed no ? I do not find it in any css/scss file

File: ui/main/src/tests/userCardView/helpers.ts [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * SPDX-License-Identifier: MPL-2.0
+ * This file is part of the OperatorFabric project.
+ */
+
+import {Process, State} from '@ofModel/processes.model';
+
+export function getProcessConfigWith(nbProcess: number, nbStatePerProcess: number) {
```
Comment (vlo-rte) [inline ui/main/src/tests/userCardView/helpers.ts]: Can you add the return type of the function please ?

**Review by vlo-rte — CHANGES_REQUESTED** [review]
Please see my remarks.

File: ui/main/src/app/business/builtInTemplates/message-or-question-list/usercard/message-or-question-listUserCardTemplateView.ts [hunk 36]
```diff
@@ -68,7 +67,7 @@ export class MessageOrQuestionListUserCardTemplateView {
         if (this.selectedMessage.question) {
             entitiesAllowedToRespond = this.selectedMessage?.possibleRecipients
                 ? this.selectedMessage?.possibleRecipients
-                : '';
+                : undefined;
```
Comment (ClementBouvierN) [inline ui/main/src/app/business/builtInTemplates/message-or-question-list/usercard/message-or-question-listUserCardTemplateView.ts]: If you want to keep undefined as the default value, then the following line is sufficient :

`entitiesAllowedToRespond = this.selectedMessage?.possibleRecipients`

File: ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts [hunk 37]
```diff
@@ -0,0 +1,614 @@
...
+            expect(card.entityRecipients).toEqual(['ENTITY1', 'ENTITY2']);
+        });
+        it('Should be set from template if field not visible', async () => {
+            setSpecificCardInformation({valid: true, card: {entityRecipients: ['ENTITY3', 'ENTITY4']}});
+            cardBuilder.setFieldVisible(InputFieldName.Recipients, false);
+            cardBuilder.setRecipientsSelectedByUser([]);
+            const card = await cardBuilder.getCard();
+            expect(card.entityRecipients).toEqual(['ENTITY3', 'ENTITY4']);
+        });
+        it('Should be the union of recipients from template and recipients form user if field is visible', async () => {
```
Comment (ClementBouvierN) [inline ui/main/src/app/business/view/userCard/cardBuilder/cardBuilder.spec.ts]: and recipients f**ro**m user

File: ui/main/src/app/business/view/userCard/cardSender/cardSender.ts [hunk 38]
```diff
@@ -0,0 +1,91 @@
...
+        if (childCard) {
+            const responseFromChildCardPost = await this.sendChildCard(
+                childCard,
+                responseFromCardPost,
+                setCurrentDateForStartDate
+            );
+            if (responseFromChildCardPost.status !== ServerResponseStatus.OK) {
+                this.displayErrorMessageOnUI();
+                logger.error(
+                    `Error while sending child card to the back end , status = ${responseFromChildCardPost.status} message =  ${responseFromChildCardPost.statusMessage}`
```
Comment (ClementBouvierN) [inline ui/main/src/app/business/view/userCard/cardSender/cardSender.ts]: "back end ," remove the space after "end"

File: ui/main/src/app/business/view/userCard/cardSender/cardSender.ts [hunk 39]
```diff
@@ -0,0 +1,91 @@
...
+                startDate: new Date().valueOf()
+            };
+        }
+        SoundNotificationService.lastSentCard(card.process + '.' + card.processInstanceId);
+        SystemNotificationService.lastSentCard(card.process + '.' + card.processInstanceId);
+        const responseFromCardPost = await firstValueFrom(CardService.postCard(cardForPublish));
+        if (responseFromCardPost.status !== ServerResponseStatus.OK) {
+            this.displayErrorMessageOnUI();
+            logger.error(
+                `Error while sending card to the back end , status = ${responseFromCardPost.status} message =  ${responseFromCardPost.statusMessage}`
```
Comment (ClementBouvierN) [inline ui/main/src/app/business/view/userCard/cardSender/cardSender.ts]: "back end ," remove the space after "end"

File: ui/main/src/app/business/view/userCard/userCard.view.ts [hunk 40]
```diff
@@ -0,0 +1,267 @@
...
+
+    public async userClickOnProcess(processId: string) {
+        if (processId === this.currentProcessId || processId == null || processId === '') return;
+        this.processStatesForm.userClickOnProcess(processId);
+        await this.initUserCardFieldsAndRenderTemplate();
+    }
+
+    public async userClickOnProcessGroup(processGroupId: string) {
+        if (processGroupId === this.processStatesForm.getSelectedProcessGroupId()) return;
+        this.processStatesForm.userClickOnProcessGroup(processGroupId);
```
Comment (ClementBouvierN) [inline ui/main/src/app/business/view/userCard/userCard.view.ts]: userClick**s**On for all these instances

File: ui/main/src/assets/i18n/en.json [hunk 41]
```diff
@@ -387,7 +387,6 @@
         "editTitle": "EDIT A NEW CARD",
         "error": {
             "notAvailable": "You are not allowed to send card.",
```
Comment (ClementBouvierN) [inline ui/main/src/assets/i18n/en.json]: Use the opportunity to change this translation to   "notAvailable": "You are not allowed to send card**s**."

It also needs to be changed in tests

File: ui/main/src/app/modules/usercard/usercard.component.html (hunk 34 above)
Comment (freddidierRTE) [reply ui/main/src/app/modules/usercard/usercard.component.html]: i think i copy/paste the code before it has been removed

File: ui/main/src/app/business/view/userCard/userCard.view.spec.ts (hunk 26 above)
Comment (freddidierRTE) [reply ui/main/src/app/business/view/userCard/userCard.view.spec.ts:52]: When you have the async keywork , the return type is a Promise 
If it is a Promise\<void\> you do not need to make return

File: ui/main/src/app/business/view/userCard/userCard.view.ts (hunk 33 above)
Comment (freddidierRTE) [reply ui/main/src/app/business/view/userCard/userCard.view.ts:209]: When you have the async keywork , the return type is a Promise If it is a Promise\<void\> you do not need to make return

**Comment by freddidierRTE** [comment]
Regarding all the exceptions , i did not manage locally to reproduce but i think i found why you have these exceptions , i pushed a new version and i will let you try on your computer if it solves the problem.

I think the problem arise when the mutliselect is initialize before the initialization of the Form classes in the view . In fact it has no impact except error in logs . To solve it i added the initialization of the forms object on the view constructor : 

```
    constructor(private userCardUIControl: UserCardUIControl) {
        this.processStatesForm = new ProcessStatesForm(this.userCardUIControl);
        this.publisherForm = new PublisherForm(this.userCardUIControl);
        this.severityForm = new SeverityForm(this.userCardUIControl);
    }
```

**Comment by freddidierRTE** [comment]
> image If it is possible, I suggest to set void for all the functions or to remove it from all of them, so that the code is homogeneous ?

Ok , I removed the void

File: ui/main/src/app/business/view/userCard/userCard.view.ts [hunk 42]
```diff
@@ -0,0 +1,274 @@
...
+        this.datesForm.userSetsDate(InputFieldName.EndDate, endDate);
+    }
+    public userSetLttd(lttd: number) {
+        this.datesForm.userSetsDate(InputFieldName.Lttd, lttd);
+    }
+    public userSetExpirationDate(expirationDate: number) {
+        this.datesForm.userSetsDate(InputFieldName.ExpirationDate, expirationDate);
+    }
+    public userSetPublisher(publisherEntityId: string) {
+        console.log('publisher selected = ', publisherEntityId);
```
Comment (vlo-rte) [inline ui/main/src/app/business/view/userCard/userCard.view.ts]: I think this log should be removed, right ?

**Review by vlo-rte — APPROVED** [review]

**Pull request merged**